### PR TITLE
[Support] Add KnownBits::computeForSubBorrow

### LINKED ISSUE #65893: [Support] Add KnownBits::computeForSubBorrow
Add the sibling to KnownBits::computeForAddCarry:

- [ ] Actual KnownBits implementation is a wrapper to the internal computeForAddCarry, similar to KnownBits::computeForAddCarry, with minor bit twiddling/swapping.
- [ ] Add a SubBorrowExhaustive unit test to KnownBitsTest.cpp
- [ ] Hook up the USUBO/SSUBO/USUBO_CARRY/SSUBO_CARRY/SUBC cases in SelectionDAG::computeKnownBits to make use of it (similar to the ADD variants already do), with suitable testing.

### PR DESCRIPTION
**Pull request by christiankissig**
- [Support] Add KnownBits::computeForSubBorrow
- [CodeGen] Implement USUBC, USUBO_CARRY, and SSUBO_CARRY with KnownBits::computeForSubBorrow
- [CodeGen] Compute unknown bits for Carry/Borrow for ADD/SUB
- [CodeGen] Compute known bits of Carry/Borrow for UADDO, SADDO, USUBO, and SSUBO

Fixes #65893

### CODE REVIEW

File: llvm/unittests/Support/KnownBitsTest.cpp
Comment (RKSimon) [inline llvm/unittests/Support/KnownBitsTest.cpp:221]: addition -> subtraction

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (nikic) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:3740]: Should be just trunc.

**Review by nikic — COMMENTED** [review]
Looks fine to me.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:3742]: nit: braces around else block if braces around if block.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:3774]: Likewise just `trunc` here.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:3775]: likewise else style.

**Comment by goldsteinn** [comment]
Might be nice to add a test where this actually shows up in CodeGen.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (christiankissig) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:3740]: @nikic and @goldsteinn Probably just hypothetical, but I couldn't work out whether Operand 2 is always of non-zero bit width, so went with the original suggestion for ADD ops to use zextOrTrunc(). Will change to trunc() on both of your recommendation here.

**Comment by christiankissig** [comment]
> Might be nice to add a test where this actually shows up in CodeGen.

+1 Will look to add it with this change.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (nikic) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:3740]: There are no zero-bit integers in SDAG (or IR). The use of zextOrTrunc in the comment probably dates back to the time when APInt did not allow zext/sext/trunc to the original bit width.

File: llvm/lib/Support/KnownBits.cpp
Comment (shafik) [inline llvm/lib/Support/KnownBits.cpp:97]:
```suggestion
                              /*CarryZero=*/Borrow.One.getBoolValue(),
                              /*CarryOne=*/Borrow.Zero.getBoolValue());
```

Comment (shafik) [reply llvm/lib/Support/KnownBits.cpp:97]: To be consistent with [bugprone-argument-comment](https://clang.llvm.org/extra/clang-tidy/checks/bugprone/argument-comment.html)

**Review by shafik — COMMENTED** [review]
Nitpick

**Comment by dlumma** [comment]
It seems like this PR is ready to land. Any reason why it has not been integrated @christiankissig ?

**Comment by christiankissig** [comment]
> It seems like this PR is ready to land. Any reason why it has not been integrated @christiankissig ?

The PR had not been merged to add a CodeGen regression test as I said I would add in https://github.com/llvm/llvm-project/pull/67788#issuecomment-1741238270 . Having looked into adding one, I realize this may need more work, compiler intrinsics for add.sub with carry. This is taking too long and may exceed the scope of the issue, so will break out into a separate piece.

I've now made the final requested changes, and this PR is ready to merge.

File: llvm/unittests/Support/KnownBitsTest.cpp
Comment (christiankissig) [reply llvm/unittests/Support/KnownBitsTest.cpp]: Done

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (christiankissig) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp]: I've replaced zextOrTrunc with trunc.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (christiankissig) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp]: Done

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (christiankissig) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp]: Done

File: llvm/include/llvm/Support/KnownBits.h
Comment (christiankissig) [reply llvm/include/llvm/Support/KnownBits.h]: Done

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Comment by goldsteinn** [comment]
Still think a test where this actually comes up in CodeGen somewhere would be nice, but the code all looks good.

**Comment by christiankissig** [comment]
> Still think a test where this actually comes up in CodeGen somewhere would be nice, but the code all looks good.

@goldsteinn I agree, this would be nice to have, but I wonder if it's worth blocking this PR on. Based on what I understand - this is my first contribution to LLVM - , we're still missing compiler intrinsics and Target lowering code, before we can test the compiler optimization on LLIR, which would all be out of scope of the ticket this is addressing. I've added regression tests ensuring that the KnownBits calculation works, so I don't think we're particularly lacking test coverage here. What are your concerns specifically?

**Comment by RKSimon** [comment]
Happy for this to go in as is (with the UADDO_CARRY -> USUBO_CARRY typo fix), this is still my preferred option.

But also happy if there's still resistance to drop the SelectionDAG changes and we just get the KnownBits helper added for now (still with exhaustive tests).

**Comment by goldsteinn** [comment]
> > Still think a test where this actually comes up in CodeGen somewhere would be nice, but the code all looks good.
> 
> @goldsteinn I agree, this would be nice to have, but I wonder if it's worth blocking this PR on. Based on what I understand - this is my first contribution to LLVM - , we're still missing compiler intrinsics and Target lowering code, before we can test the compiler optimization on LLIR, which would all be out of scope of the ticket this is addressing. I've added regression tests ensuring that the KnownBits calculation works, so I don't think we're particularly lacking test coverage here. What are your concerns specifically?

Sorry missed this, I didn't mean for the concern to be blocking this.

If its not really doable that okay and this change LGTM.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (christiankissig) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp]: Done

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by RKSimon** [comment]
@christiankissig Do you have commit access yet or shall I squash + merge it?

**Comment by christiankissig** [comment]
I don't have access yet. Please merge. Thank you!

**Pull request merged**
